File: util/src/test/java/com/tars/util/OptionTest.java
package com.tars.util;

import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

public class OptionTest {

  @Test
  public void testOf() {
    assertEquals(Option.of("value").get(), "value");
  }

  @Test(expected = NullPointerException.class)
  public void testOfIllegal() {
    Option.of(null);
  }

  @Test
  public void testOfNullable() {
    Option<String> opt = Option.ofNullable("value");

    assertNotNull(opt);
    assertEquals(opt.get(), "value");

    Option.ofNullable(null);
  }

  @Test
  public void testGet() {
    assert Option.of("value").get().equals("value");
  }

  @Test
  public void testIsPresent() {
    assert Option.of("value").isPresent();
  }

  @Test
  public void testIfPresent() {
    Option.of("value")
        .ifPresent((value) -> assertTrue(true))
        .ifNotPresentThrow(IllegalStateException::new);
  }

  @Test
  public void testFilter() {
    Option.of("some text")
        .filter(s -> s.startsWith("some"))
        .ifNotPresentThrow(IllegalStateException::new);
  }

  @Test
  public void testMap() {
    Option.of("")
        .map(String::hashCode)
        .ifPresent((value) -> assertTrue(value == 0))
        .ifNotPresentThrow(IllegalStateException::new);
  }

  @Test
  public void testFlatMap() {
    Option.of("value")
        .flatMap(s -> Option.of(s.length()))
        .ifPresent((value) -> assertTrue(value == 5))
        .ifNotPresentThrow(IllegalStateException::new);
  }

  @Test
  public void testOrElse() {
    Option.empty()
        .ifNotPresentTake("value")
        .ifPresent((value) -> assertEquals(value, "value"))
        .ifNotPresentThrow(IllegalStateException::new);
  }

  @Test
  public void testOrElseGet() {
    Option.empty()
        .ifNotPresentTake(() -> "value")
        .ifPresent((value) -> assertEquals(value, "value"))
        .ifNotPresentThrow(IllegalStateException::new);

  }

  @Test(expected = IllegalStateException.class)
  public void testOrElseThrow() {
    Option.empty()
        .ifNotPresentThrow(IllegalStateException::new);
  }

  @Test
  public void testEquals() {
    assert Option.of("value").equals(Option.of("value"));
    assert Option.ofNullable(null).equals(Option.empty());
  }

  @Test
  public void testHashCode() {
    assert Option.empty().hashCode() == 0;
  }

  @Test
  public void testToString() {
    assert Option.of("value").toString().equals("Option[value]");
  }

  public static void main(String[] args) {
//    long millisecondsSinceEpoch = System.currentTimeMillis();
//
//    DateTimeFormatter.ofLocalizedTime (FormatStyle.SHORT).format(Instant.ofEpochMilli (millisecondsSinceEpoch ));
//
//
//
//    DateTimeFormatter formatter =
//    String output = formatter.format ( zdt );
//
//    System.out.println ( "millisecondsSinceEpoch: " + millisecondsSinceEpoch + " instant: " + instant + " output: " + output );
  }
}

File: sources/global/util/src/test/kotlin/util/common/OptionTest.kt
package util.common

import org.junit.Test
import util.Option
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class OptionTest {
    @Test(expected = IllegalStateException::class)
    fun testGetEmpty() {
        Option.empty.get()
    }

    @Test
    fun testIsPresent() {
        assertTrue { Option.Companion.of(1).isPresent() }
    }

    @Test
    fun testGetExisting() {
        assertEquals("val", Option.Companion.of("val").get())
    }

    @Test
    fun testFilter() {
        assertEquals(
                Option.Companion.empty(),
                Option.Companion.of("val").filter { it.isEmpty() }
        )
    }

    @Test
    fun testMap() {
        assertEquals(
                Option.Companion.of("VAL"),
                Option.Companion.of("val").map { it.toUpperCase() })
    }

    @Test
    fun testFlatMap() {
        assertEquals(
                Option.Companion.of("VAL"),
                Option.Companion.of("val").flatMap { Option.Companion.of(it.toUpperCase()) }
        )
    }

    @Test
    fun testIfNotPresentTakeValue() {
        assertEquals(
                Option.Companion.of("val"),
                Option.Companion.empty<String>().ifNotPresentTake("val")
        )
    }

    @Test
    fun testIfNotPresentTakeSupplier() {
        assertEquals(
                Option.Companion.of("val"),
                Option.Companion.empty<String>().ifNotPresentCompute { "val" }
        )
    }

    @Test
    fun testTake() {
        assertEquals(
                Option.Companion.of(1),
                Option.Companion.of("val").take(1)
        )
        assertEquals(
                Option.Companion.of(1),
                Option.Companion.of("val").take(Option.Companion.of(1))
        )
    }

    @Test
    fun testIfPresentAction() {
        var ifPresentExecuted = false
        var ifNotPresentExecuted = false

        Option.of("val")
                .ifPresent { ifPresentExecuted = true }
                .ifNotPresent { ifNotPresentExecuted = true }

        assertTrue { ifPresentExecuted }
        assertFalse { ifNotPresentExecuted }
    }

    @Test
    fun testIfNotPresentAction() {
        var ifPresentExecuted = false
        var ifNotPresentExecuted = false

        Option.empty
                .ifPresent { ifPresentExecuted = true }
                .ifNotPresent { ifNotPresentExecuted = true }

        assertFalse { ifPresentExecuted }
        assertTrue { ifNotPresentExecuted }
    }

    @Test
    fun testClear() {
        assertEquals(
                Option.Companion.empty,
                Option.Companion.of("val").clear()
        )
    }

    @Test
    fun sample(){


    }

    private fun aToB(it: Int): String {
        throw UnsupportedOperationException("not implemented")
    }

}
